include/graph.h

```diff
@@ -0,0 +1,46 @@
+#pragma once
+
+#include "types.h"
+
+#include <memory>
+#include <vector>
+
+namespace celerity::detail {
+
+/// A `graph` keeps ownership of all graph nodes that have not been pruned by epoch or horizon application.
+template <typename Node, typename NodeIdLess>
+class epoch_partitioned_graph {
+	friend struct graph_testspy;
+
+  public:
+	// Call this before pushing horizon or epoch node in order to be able to call erase_before_epoch on the same task id later.
+	void begin_epoch(const task_id tid) {
+		assert(m_epochs.empty() || (m_epochs.back().epoch_tid < tid && !m_epochs.back().nodes.empty()));
+		m_epochs.push_back({tid, {}});
```

Comment:
For the TDAG it would make sense to set this to the horizon step size, but for IDAG the choice is somewhat arbitrary and I suggest we leave it up to the (C++) implementation.

Reply:
AFAIK all implementations start with 0, but let's keep it as-is for now.
(I do think horizon step size as a minimum would never really be wrong though -- we can revisit if there is any performance reduction)

Reply:
Starting at 0 is fine, the important question is how large the allocation made by the first `push_back` is.